src/lib.rs

```diff
@@ -699,4 +824,99 @@ mod unix {
             .try_into()
             .context("Failed to get page size")
     }
+
+    impl PageFaultChecker {
+        pub(super) fn new(ptr: *mut usize, len: usize) -> PageFaultChecker {
+            PageFaultChecker {
+                ptr,
+                len,
+                state: None,
+            }
+        }
+    }
+
+    unsafe impl Send for PageFaultChecker {}
+
+    impl super::memtest::TestObserver for PageFaultChecker {
+        type Error = PageFaultError;
+
+        // TODO: Currently NUM_CHECKS is an arbitrary number
+        fn init(&mut self, expected_iter: u64) {
+            const FIRST_CHECKPOINT: u64 = 0;
+            const NUM_CHECKS: u64 = 100;
+
+            assert!(
+                self.state.is_none(),
+                "init() should only be called once per test"
+            );
+
+            self.state = Some(PageFaultCheckerState {
+                completed_iter: 0,
+                checkpoint: FIRST_CHECKPOINT,
+                checkpoint_step: expected_iter / NUM_CHECKS,
+            });
+        }
+
+        #[inline(always)]
+        fn check(&mut self) -> Result<(), Self::Error> {
+            let state = self
+                .state
+                .as_mut()
+                .expect("init() should be called before check()");
+
+            if state.completed_iter < state.checkpoint {
+                state.completed_iter += 1;
+                return Ok(());
+            }
+
+            state.on_checkpoint(self.ptr, self.len)
+        }
+    }
+
+    impl PageFaultCheckerState {
+        fn on_checkpoint(&mut self, ptr: *mut usize, len: usize) -> Result<(), PageFaultError> {
+            use libc::mincore;
+
+            unsafe {
+                let page_size = get_page_size().unwrap();
+                let base_address = ptr as usize;
+                let mem_size = len * size_of::<usize>();
+
+                let aligned_address = base_address & !(page_size - 1);
```

Comment:
I'd say the more straightforward and canonical way to do this is simply:
```Rust
let aligned_address = base_address / page_size * page_size;
```

That idiom is used so often that virtually every compiler has the optimization for it hard-coded in.